### PR TITLE
Use faster globs for template resolving

### LINKED ISSUE #20752: ActionView template finder can become really slow due to disk IO
As the number of view paths, template handlers and formats increases.

Caching `ActionView::PathResolver#find_template_paths` as follows sped the rendering of an API call (rabl/json) **from 2500ms to 30ms** once the cache was warmed up:

``` ruby
ActionView::PathResolver.class_eval do
  def find_template_paths(query)
    Rails.cache.fetch([:action_view_path_resolver_cache, query]) do
      Dir[query].reject { |filename|
        File.directory?(filename) ||
          # deals with case-insensitive file systems.
          !File.fnmatch(query, filename, File::FNM_EXTGLOB)
      }
    end
  end
end
```
### Details

The app is Rails v4.1.11 / Ruby 2.2.1 / Spree 2-4-stable. I also noticed really slow performance for partial lookups relying on controller inheritance (e.g. `render 'header'` where `_header.html` is in `app/views/application`), I think it might be for the same reason.

I got this results on a Spree 2.4 app with Rails 4.1.11 calling [this API endpoint](https://github.com/spree/spree/blob/2-4-stable/api/app/controllers/spree/api/states_controller.rb#L8-L20) during which `#find_template_paths` [is called 137 times](https://gist.github.com/elia/e3e624e2235f12946ea4) with queries as long as:

```
/Users/elia/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/kaminari-0.16.3/app/views/layouts/action_controller/base{.en,}{.json,}{.v10,.v9,.v8,.v7,.v6,.v5,.v4,.v3,.v2,.v1,}{}{.erb,.builder,.raw,.ruby,.prawn,.coffee,.opal,.haml,.rabl,.jbuilder,}
```

From what I saw not much has changed on rails@master wrt the path lookup, but I could be wrong. I agree that this can be regarded as a Spree issue but I believe that similar problems arise in many Rails apps.

_just for completeness the logs of the two requests follow_

Without the patch:

```
Started GET "/api/states.json?country_id=155" for 127.0.0.1 at 2015-07-01 23:49:58 +0200
Processing by Spree::Api::StatesController#index as JSON
  Parameters: {"country_id"=>"155"}
  Spree::User Load (0.5ms)  SELECT  "spree_users".* FROM "spree_users"  WHERE "spree_users"."deleted_at" IS NULL AND "spree_users"."spree_api_key" = '' LIMIT 1
  Spree::Country Load (0.3ms)  SELECT  "spree_countries".* FROM "spree_countries"  WHERE "spree_countries"."id" = $1 LIMIT 1  [["id", 155]]
  Spree::State Load (0.2ms)  SELECT  "spree_states".* FROM "spree_states"  WHERE "spree_states"."country_id" = $1  ORDER BY name DESC LIMIT 1  [["country_id", 155]]
  Spree::State Load (0.4ms)  SELECT "spree_states".* FROM "spree_states"  WHERE "spree_states"."country_id" = $1  ORDER BY name ASC  [["country_id", 155]]
  Rendered /Users/elia/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/bundler/gems/spree-aa554d622f47/api/app/views/spree/api/states/index.v1.rabl (1.7ms)
Completed 200 OK in 2848ms (Views: 2829.6ms | ActiveRecord: 1.9ms)
```

With the patch:

```
Started GET "/api/states.json?country_id=155" for 127.0.0.1 at 2015-07-01 23:49:05 +0200
Processing by Spree::Api::StatesController#index as JSON
  Parameters: {"country_id"=>"155"}
  Spree::User Load (0.7ms)  SELECT  "spree_users".* FROM "spree_users"  WHERE "spree_users"."deleted_at" IS NULL AND "spree_users"."spree_api_key" = '' LIMIT 1
  Spree::Country Load (0.3ms)  SELECT  "spree_countries".* FROM "spree_countries"  WHERE "spree_countries"."id" = $1 LIMIT 1  [["id", 155]]
  Spree::State Load (0.2ms)  SELECT  "spree_states".* FROM "spree_states"  WHERE "spree_states"."country_id" = $1  ORDER BY name DESC LIMIT 1  [["country_id", 155]]
  Spree::State Load (0.4ms)  SELECT "spree_states".* FROM "spree_states"  WHERE "spree_states"."country_id" = $1  ORDER BY name ASC  [["country_id", 155]]
  Rendered /Users/elia/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/bundler/gems/spree-aa554d622f47/api/app/views/spree/api/states/index.v1.rabl (1.8ms)
Completed 200 OK in 31ms (Views: 7.7ms | ActiveRecord: 2.0ms)
```
### Solution (?)

The above patch of course is not good (even in concept) for development where views are added and modified regularly. A possible solution could be to keep the lookup dynamic just inside the app root and cache the paths coming from gems. That in turn could be an issue for people relying on the fact that views are reloaded even for gems in development (e.g. pleople using engines) but to me seems a good trade off anyway.

cc @tenderlove @masterkain

### PR DESCRIPTION
**Pull request by jhawthorn**
Fixes #20752

`ActionView::PathResolver` (including the more optimized subclasses) works by evaluating a large glob pattern.

Its search is mostly equivalent to:
```
Dir["app/views/users/show{.en,}{.html,.text,.js,.css,.ics,.csv,.vcf,.vtt,.png,.jpeg,.gif,.bmp,.tiff,.svg,.mpeg,.mp3,.ogg,.m4a,.webm,.mp4,.otf,.ttf,.woff,.woff2,.xml,.rss,.atom,.yaml,.multipart_form,.url_encoded_form,.json,.pdf,.zip,.gzip,}{}{.raw,.erb,.html,.builder,.ruby,}"]
```

This is relatively slow. Ruby will check the existance of each possible file:
First `app/views/users/show.en.html.raw`,
then `app/views/users/show.en.html.erb`,
then `app/views/users/show.en.html.html`, ....etc

For the example above, which is not the worst - variants and fallback locales would add to complexity, that's 2 * 35 * 6 = 420 checks. Not nice.

Each of these checks needs 4 syscalls (it checks each path up the tree), resulting in [**1682 syscalls**](https://gist.github.com/jhawthorn/d9410cdf835d9229507450dc8dd45059). This would be even worse in a real app because it scans an absolute path.

It's much faster, as proposed in #20752 to scan the directory with a wildcard, and then filter the results in Ruby.

```
Dir["app/views/users/show*"]
```

This is only [**14 syscalls**](https://gist.github.com/jhawthorn/bc6f08db5e952267dec9fa2b976dda60). Great!

---

Filtering this would be trivial using the existing glob and `File.fnmatch`, but the problem we then run into is sorting (there is some discussion of this in #20752).

Although Ruby's docs don't guarantee an order returned by `Dir.glob`, it is consistent, and Rails relies on that order (ex. we want to prefer `show.en.html.erb` to `show.html.erb`).

This PR solves this by also building a regex to match paths and then using it to both ensure each file actually matches (vs just shares the prefix) and extract the details values which were matched. We then use the matched values to sort the results into the same order they would have been returned by the original glob pattern.

## Benchmark

```
require './config/environment'

Benchmark.ips do |x|
  x.report "lookup users/show" do
    ApplicationController.new.lookup_context.find_template("users/show", [], false, {}, {})
  end
end
```

**Before**

```
Warming up --------------------------------------
   lookup users/show     6.000  i/100ms
Calculating -------------------------------------
   lookup users/show     61.302  (± 4.9%) i/s -    306.000  in   5.011577s
```

**After**

```
Warming up --------------------------------------
   lookup users/show   353.000  i/100ms
Calculating -------------------------------------
   lookup users/show      3.845k (± 5.3%) i/s -     19.415k in   5.064962s
```

---

Thanks for reading ❤️

cc @eileencodes @brianmario @tenderlove

### CODE REVIEW

**Comment by rafaelfranca** [comment]
Build is broken.

**Comment by Erowlin** [comment]
Should this be back-ported to 5.X ?

**Comment by clemensp** [comment]
@jhawthorn : 
Thanks for the fix. Addressed slowdowns my team was experiencing after extracting a bunch of modules with rails engines.
I'm eager to get this back-ported, and am willing to support the efforts to make it happen.

@rafaelfranca : Which versions of Rails would accept this back-port? I'm interested in getting this into 4.2, 5.0, and 5.1. I did some searching, but couldn't find the policy around back-porting.

Thanks!

**Comment by jeremy** [comment]
@clemens Check out the [Maintenance Policy](https://guides.rubyonrails.org/maintenance_policy.html) for what gets backported 👍

**Comment by clemensp** [comment]
@jeremy : Thanks. I guess based on the Maintenance Policy, a performance optimization mainly for development mode would be categorized under "Bug Fixes" at best.

Is the Rails team interested in a PR for Rails 5.2?

Also, any chance a back-port for Rails 5.0 or 5.1 would be accepted?
My team's going through a long upgrade process to get to the latest Rails version, and it would be great to get the fix back-ported instead of having to maintain a custom back-port. I support whatever decision the Rails team makes. :)

File: actionview/lib/action_view/template/resolver.rb [hunk 1]
```diff
@@ -362,19 +365,56 @@ def eql?(resolver)
...
-        else
-          "{#{details[ext].compact.uniq.map { |e| "#{prefix}#{e}," }.join}}"
+      def find_template_paths_from_details(path, details)
+        # Instead of checking for every possible path, as our other globs would
+        # do, scan the directory for files with the right prefix.
+        query = "#{escape_entry(File.join(@path, path))}*"
+
+        regex = build_regex(path, details)
+
+        Dir[query].uniq.reject do |filename|
```
Comment (semaperepelitsa) [inline actionview/lib/action_view/template/resolver.rb]: When does Dir.[] return duplicate file paths?

**Pull request merged**